--- a/Eigen/src/Core/util/Constants.h
+++ b/Eigen/src/Core/util/Constants.h
@@ -22,21 +22,21 @@ namespace Eigen {
  *
  * Changing the value of Dynamic breaks the ABI, as Dynamic is often used as a template parameter for Matrix.
  */
-const int Dynamic = -1;
+inline constexpr int Dynamic = -1;
 
 /** This value means that a signed quantity (e.g., a signed index) is not known at compile-time, and that instead its
  * value has to be specified at runtime.
  */
-const int DynamicIndex = 0xffffff;
+inline constexpr int DynamicIndex = 0xffffff;
 
 /** This value means that the requested value is not defined.
  */
-const int Undefined = 0xfffffe;
+inline constexpr int Undefined = 0xfffffe;
 
 /** This value means +Infinity; it is currently used only as the p parameter to MatrixBase::lpNorm<int>().
  * The value Infinity there means the L-infinity norm.
  */
-const int Infinity = -1;
+inline constexpr int Infinity = -1;
 
 /** This value means that the cost to evaluate an expression coefficient is either very expensive or
  * cannot be known at compile time.
@@ -45,7 +45,7 @@ const int Infinity = -1;
  * and very very expensive expressions. It thus must also be large enough to make sure unrolling won't happen and that
  * sub expressions will be evaluated, but not too large to avoid overflow.
  */
-const int HugeCost = 10000;
+inline constexpr int HugeCost = 10000;
 
 /** \defgroup flags Flags
  * \ingroup Core_Module
@@ -67,16 +67,16 @@ const int HugeCost = 10000;
  * For an expression, this determines the storage order of
  * the matrix created by evaluation of that expression.
  * \sa \blank  \ref TopicStorageOrders */
-const unsigned int RowMajorBit = 0x1;
+inline constexpr unsigned int RowMajorBit = 0x1;
 
 /** \ingroup flags
  * means the expression should be evaluated by the calling expression */
-const unsigned int EvalBeforeNestingBit = 0x2;
+inline constexpr unsigned int EvalBeforeNestingBit = 0x2;
 
 /** \ingroup flags
  * \deprecated
  * means the expression should be evaluated before any assignment */
-EIGEN_DEPRECATED const unsigned int EvalBeforeAssigningBit = 0x4;  // FIXME deprecated
+EIGEN_DEPRECATED inline constexpr unsigned int EvalBeforeAssigningBit = 0x4;  // FIXME deprecated
 
 /** \ingroup flags
  *
@@ -94,7 +94,7 @@ EIGEN_DEPRECATED const unsigned int EvalBeforeAssigningBit = 0x4;  // FIXME depr
  * \note This bit can be set regardless of whether vectorization is actually enabled.
  *       To check for actual vectorizability, see \a ActualPacketAccessBit.
  */
-const unsigned int PacketAccessBit = 0x8;
+inline constexpr unsigned int PacketAccessBit = 0x8;
 
 #ifdef EIGEN_VECTORIZE
 /** \ingroup flags
@@ -105,9 +105,9 @@ const unsigned int PacketAccessBit = 0x8;
  * If vectorization is not enabled (EIGEN_VECTORIZE is not defined) this constant
  * is set to the value 0.
  */
-const unsigned int ActualPacketAccessBit = PacketAccessBit;
+inline constexpr unsigned int ActualPacketAccessBit = PacketAccessBit;
 #else
-const unsigned int ActualPacketAccessBit = 0x0;
+inline constexpr unsigned int ActualPacketAccessBit = 0x0;
 #endif
 
 /** \ingroup flags
@@ -130,7 +130,7 @@ const unsigned int ActualPacketAccessBit = 0x0;
  * Product is a vector expression. Thus, vector Product expressions allow index-based coefficient access but
  * not index-based packet access, so they don't have the LinearAccessBit.
  */
-const unsigned int LinearAccessBit = 0x10;
+inline constexpr unsigned int LinearAccessBit = 0x10;
 
 /** \ingroup flags
  *
@@ -145,7 +145,7 @@ const unsigned int LinearAccessBit = 0x10;
  * Expressions having LvalueBit also have their coeff() method returning a const reference instead of returning a new
  * value.
  */
-const unsigned int LvalueBit = 0x20;
+inline constexpr unsigned int LvalueBit = 0x20;
 
 /** \ingroup flags
  *
@@ -156,7 +156,7 @@ const unsigned int LvalueBit = 0x20;
  *
  * See the comment on LvalueBit for an explanation of how LvalueBit and DirectAccessBit are mutually orthogonal.
  */
-const unsigned int DirectAccessBit = 0x40;
+inline constexpr unsigned int DirectAccessBit = 0x40;
 
 /** \deprecated \ingroup flags
  *
@@ -168,9 +168,9 @@ const unsigned int DirectAccessBit = 0x40;
  * expression.packet<Aligned>(0);
  * \endcode
  */
-EIGEN_DEPRECATED const unsigned int AlignedBit = 0x80;
+EIGEN_DEPRECATED inline constexpr unsigned int AlignedBit = 0x80;
 
-const unsigned int NestByRefBit = 0x100;
+inline constexpr unsigned int NestByRefBit = 0x100;
 
 /** \ingroup flags
  *
@@ -179,7 +179,7 @@ const unsigned int NestByRefBit = 0x100;
  * The precise choice will be decided at evaluation time or when
  * combined with other expressions.
  * \sa \blank  \ref RowMajorBit, \ref TopicStorageOrders */
-const unsigned int NoPreferredStorageOrderBit = 0x200;
+inline constexpr unsigned int NoPreferredStorageOrderBit = 0x200;
 
 /** \ingroup flags
   *
@@ -192,10 +192,10 @@ const unsigned int NoPreferredStorageOrderBit = 0x200;
     inline const Index* innerNonZeroPtr() const;
     \endcode
   */
-const unsigned int CompressedAccessBit = 0x400;
+inline constexpr unsigned int CompressedAccessBit = 0x400;
 
 // list of flags that are inherited by default
-const unsigned int HereditaryBits = RowMajorBit | EvalBeforeNestingBit;
+inline constexpr unsigned int HereditaryBits = RowMajorBit | EvalBeforeNestingBit;
 
 /** \defgroup enums Enumerations
  * \ingroup Core_Module
